### ISSUE TITLE
cguid not unique

### ISSUE THREAD

**Issue by pchainho** [reTHINK-project/dev-catalogue#50]
While debugging the [timeout](https://github.com/reTHINK-project/specs/issues/7) bug we've realised that [catalogue objects](https://github.com/reTHINK-project/specs/tree/master/datamodel/core/hyperty-catalogue) are not having `cguid` as global unique identifiers [causing problems when deploying catalogue objects](https://github.com/reTHINK-project/dev-service-framework/blob/revised-catalogue/src/runtime-catalogue/RuntimeCatalogue.js#L82).

**Comment by Endebert** [comment]
@pchainho could you specify which catalogue objects are interfering with each other? Or did you use custom objects in your testbed?

**Comment by pchainho** [comment]
One example is the default protostub for "hybroker.rethink.ptinovacao.pt" and "rethink.tlabscloud.com" that have the same "cguid" `40001`. Thus, in cross domain scenarios, the runtime catalogue local storage have the same key for two different protostubs.

But afaik this is a general problem that can happen to any catalogue object, not only to default protostubs.

There should be a way to avoid that, eg using the catalogue domain in the cguid.

**Comment by emmelmann-fokus** [comment]
It is the responsibility of the person providing the catalogue object to put a GUID in the field. The means on how to create such a GUID is out of scope here.

One means could be to use: https://guidgenerator.com

No need to change anything in the spec or implementation.

**Comment by pchainho** [comment]
Disagree, we can't force developers to use this. It is safer to be the catalogue ensuring we have the GUID in the field.

**Comment by emmelmann-fokus** [comment]
We are not forcing developers to use this tool.  They may use any mean they which to generate a GUID.  The catalogue is hosting correct catalogue objects.

If the catalogue were to automatically place a GUID in the field upon loading, the GUID would likely not be persistent.  Also, it is very well possible to host the same catalogue object in several catalogues. To identify the catalogue object, we need this GUID. And it cannot be generated by the catalogue as it has to be the same for all catalogue objects deployed on  a worldwide basis.

**Comment by sbecot** [comment]
A question: why do we provide GUIDs for packages (why Global)? Wouldn't an ID and domain be enough? If I take the GUID of an already online catalogue object, and put online my catalogue, who will notice that there is a problem -except when a bug occurs-? 
The global registry is the only instance managing GUIDs, and it is dedicated to users, not to catalogue objects. So nothing is planned to take care of globally unique identifiers for source objects. The catalogue can just ensure that it is providing unique id within its own scope, so adding the domain could make sense (managed by the runtime or the catalogue???).

@emmelmann-fokus "it is very well possible to host the same catalogue object in several catalogues": If so they may be considered as different objects. 
If not it will be very difficult to manage: if we consider the GUID has to be generated with something ensuring its uniqueness no doubt that we need an authoring tool, that would also generate and deploy the packages, and generate the GUID at the same time. 

Temporarily, we can find a workaround (the proposal of Marc could be one). We also could add an information to ensure uniqueness. For example add a base64 encoding of the creator. Example: 100001am9obi5kb2VAZm9vLmZy is the first hyperty created by am9obi5kb2VAZm9vLmZy alias john.doe@foo.fr . This is far from perfect, but at least for the tests it would be ok.

In a second time we need to clarify the construction of ID (maybe not GUID) of catalogue objects.

**Comment by emmelmann-fokus** [comment]
Hi,
I don't want to re-open the discussion again why we added a _G_UID to each catalogue object.  There are many reasons to do and as many not to do it.

From what I remember of the discussion we had way back at the Aveiro meeting: 
* The same catalogue object might be stored in several catalogues.  E.g., if we develop a hyperty, this hyperty might be stored in the catalogue hosted by several service providers. This yielded to the decision to have a GUID which uniquely identifies a hyperty etc. (i.e., an object stored in the catalogue).  Side note: if we add any information automatically or internally -- e.g. by concatenating the creator ID or catalogue domain name etc. -- to the catalogue object id, we creating global uniqueness; so we are not arguing about the *global* uniqueness but about the *means* how to provide the GUID.
* As there might be several versions of a catalogue object (e.g. when bugs are fixed), we aimed for a version number in addition to the GUID.

Using GUIDs is a well accepted means to identify objects without causing "id conflicts".  And considering two objects with the same id in different catalogues as different objects actually means that we introduce GUIDs (= id + catalogue-identified).

I do not see any reason to change the agreed specification of catalogue objects having GUIDs as I cannot believe that asking a developer to tag their object with a GUID is hard; if you tell a developer to generate a GUID by whatever means he wants, he should be well capable to do so, even without an authoring tool.

**Comment by pchainho** [comment]
@emmelmann-fokus the main point is that if it is not the catalogue or some other provisioning tool working together with the catalogue, there is no way to ensure the cguid is unique. 

It is not hard having this done by developers, of course, but there could always be accidents or even malicious developers that can take advantage of this weak point.

Another point is that there are source packages that could be the same but the descriptors will be different. Example, domains using the same protostub like hybroker.rethink.ptinovacao.pt and rethink.quobis.com. They are both using the same protostub implementation but the descriptor configuration is different in order to let them connect to different wss URLs

In case the developer of the vertx protostub uses the toolkit gulp task to generate the cguid, all domains using this implementation would have the same cguid in the catalogue. But then, interoperability between these domains would not work.

A quick fix would be to skip the cguid, and use the catalogue URL as the key when persisting the catalogue object in the runtime. But we should get a better solution in future.

**Comment by emmelmann-fokus** [comment]
> A quick fix would be to skip the cguid, and use the catalogue URL as the key when persisting the catalogue object in the runtime. But we should get a better solution in future.

Please open a dedicated issue before changing anything in the catalogue object specification.  The implementation might use the GUID internally for indexing / keeping track of catalogue.  The spec of the catalogue object said to have a globally unique ID in that place and we have to ensure that implementations do not break before changing specifications.

**Comment by sbecot** [comment]
I probably missed this point in Aveiro, but it seems that asking to provide a unique  global ID with no mean to generate/control it is a bit tricky or even risky. 
> It is not hard having this done by developers, of course, but there could always be accidents or even malicious developers that can take advantage of this weak point.

I fully agree. As far as I know, developers are not disciplined people if they are not constrained. And also would it mean that a checksum should be provided to ensure no code modification? If 2 codes differents have same cguid, who is taken into account?
Point taken to open an issue to clarify how to build and manage this ID

@pchainho about:
> Another point is that there are source packages that could be the same but the descriptors will be different. Example, domains using the same protostub like hybroker.rethink.ptinovacao.pt and rethink.quobis.com. They are both using the same protostub implementation but the descriptor configuration is different in order to let them connect to different wss URLs
In case the developer of the vertx protostub uses the toolkit gulp task to generate the cguid, all domains using this implementation would have the same cguid in the catalogue. But then, interoperability between these domains would not work.

It's not very clear. How would it be possible that the same code is not compatible when we claim interoperability with different codes? Could you explain more? Maybe not in this issue anyway.

**Comment by emmelmann-fokus** [comment]
Honestly, I would suggest to leave everything as it is.  Just generate within the project a GUID and use it. That will avoid any issues.

For later deployment, no matter what you do, unless you establish a "well controlled entity" (however that will look like), you will never solve the issue to counteract 
>accidents or even malicious developers that can take advantage of this weak point.

Architecturally, there are two aspects here:

1. How to generate a GUID
2. How to assure that a developer did actually generate a GUID (i.e. checking that the GUID is unique and has not been used before)
3. Assuring that there is no (intended) miss use of the GUID

Let's do it in reverse order: (ad 3) even today, I can change the MAC address of an Ethernet Card and claim to be someone else.  Hard to do but possible.  Or I can produce new Ethernet Cards (in my hidden factory in the basement :-) and use for the MAC address (which should be globally unique) the prefix that is allocated to another manufacturer.  You see, there is more than what we can do within the resources of the project to provide a ready-to-run solution for this.

 (ad 2) Simply: you can't unless you establish a synchronized worldwide database which is the master to test against if a GUID has been used before.  The algorithm to generate a GUID is to produce a "tag / number" whose probability is neglect ably small to be reproduced again when running the same algorithm.  The key here is the algorithm and usually, you cannot reengineer the algorithm based on a single instance of a produced result.  So you -- if you only see a single GUID -- can never assure that this GUID is really globally unique.

(ad 1) there are many many ways on doing this.  One is to include an "assigned prefix" (e.g. the catalogue name).  But also, there is trust here that the same name is not used twice etc.


We need "a form of GUID" which has to be persistent for an object once assigned.  I'll leave it to you to decide how to generate it.  The catalogue is expecting a GUID in the provided data base objects per object.  Just please open a dedicated issue before you change the format of the GUID in the catalogue objects as @Endebert has to give permission for such a change due to potential implementation issues.

cheers,
M

**Comment by pchainho** [comment]
@emmelmann-fokus that's why cguid can't be generated by developers but by the Catalogue. I guess the catalogue URL would work and no further sophisticated algorithms would be needed. But its up to you about the best and simplest way to generate it.

> How would it be possible that the same code is not compatible when we claim interoperability with different codes? Could you explain more? Maybe not in this issue anyway.

@sbecot this issue was raised exactly because it was causing errors in some interoperability scenarios. ie since there was the same cguid for two different domain protostubs, the msg bus was delivering messages to the wrong protostub.

**Comment by emmelmann-fokus** [comment]
Well to be honest and to the point, the errors were caused because people that deployed two different catalogue objects did not follow the specification but used both the same GUID (which was not globally unique).

If the catalogue generates the GUID for objects itself, it cannot guarantee that this GUID is persistent.  It will change each time a database is (re)loaded.  If you want a persistent GUID for each catalogue object, such GUID has to be persistently stored in the catalogue objects description (i.e. the catalogue object itself) and has to be provided by the person / developer providing this object.

To move forward here, please open a new, dedicated issue, to start discussion on the specification of the GUID.  We will need a firm statement on the constraints and characteristics of the GUID field int the catalogue object before we can effectively progress on who, where and how to provide the value for this field.

**Comment by pchainho** [comment]
@emmelmann-fokus we can postpone the final decision with changes to specs if required but we should agree on a quick fix now. That's the purpose of this issue. 

If you don't want to change the catalogue at this point, the quick fix I propose, is not to use the cguid and just the catalogue URL as the key for persisting the catalogue objects at Runtime. If you agree with this, then we can close this issue and open another issue for the Runtime Catalogue. What do you think?

**Comment by emmelmann-fokus** [comment]
@pchainho I would like to talk to Robert first.  If I remember correctly, the catalogue uses internally the GUID as an index to the stored data -- which requires it to be a GUID.  As a quick fix, I would suggest to put a GUID in the GUID field.

I have the feeling that it is architecturally wrong to use the catalogue URL to the object for persistence as different catalogues may store the identical object.  But your proposal does not hurt; it would just cause updates of already cached objects (i.e. false negatives) which is not as bad as having false positives in this particular case.

If you implement any code to avoid using the GUID, try to encapsulate such code in a dedicated function (i.e.: get_GUID() ) which returns "something" which is globally unique.  Just a suggestion to allow for easy migration of additional code changes.

**Comment by Endebert** [comment]
Maybe we should first define what exactly the GUID is supposed to represent. From my point of view, I assumed the GUID is supposed to be unique for descriptor + sourcePackage, i.e. source code and its configuration.

So. if you say:

> One example is the default protostub for "hybroker.rethink.ptinovacao.pt" and "rethink.tlabscloud.com" that have the same "cguid" 40001.

then it implies that the default protostub for both are exactly the same, i.e source code + configuration.

> Thus, in cross domain scenarios, the runtime catalogue local storage have the same key for two different protostubs.

But, according to what I stated above, both protostubs are supposed to be _identical_. So it should not matter, from where this protostub was retrieved initially.

If this is not the case, then the person that reused an already "taken" cguid is at fault. The solution: change the cguid for one of them.

Until we have a solution that ensures a unique cguid on a global scale, you have to be careful not to reuse existing cguids.

From an implementation standpoint, the Catalogue Broker/Dabasase does not use GUID internally any more since version 1.3.0, so you can set/change it to whatever you like (before that, the cguid was used to map the sourcePackage to a descriptor).

> There should be a way to avoid that, eg using the catalogue domain in the cguid.

Since the cguid field is handled as a string, you can change the value by any means you see fit.
So when deploying the descriptors in the Catalogue Database, feel free to add a domain to the cguid fields.
However, the Catalogue itself will not change them automatically.

In my opinion, there is currently no need to make a change in the Runtime Catalogue to solve this problem. Instead, the differing descriptors (in this case protostubs) just need differing cguids.

**Comment by pchainho** [comment]
> then it implies that the default protostub for both are exactly the same, i.e source code + configuration.

They are not. The configuration from the different catalogue objects can be different and are not set by the developers but by the service provider. And each service provider is not able to know what are the "cguid" used by other service providers.

> In my opinion, there is currently no need to make a change in the Runtime Catalogue to solve this problem. Instead, the differing descriptors (in this case protostubs) just need differing cguids

It is clear from this discussion that currently "cguid" is a problem not a solution and should not be used in the runtime.

**Comment by emmelmann-fokus** [comment]
Thanks all for the clarifications.

> From an implementation standpoint, the Catalogue Broker/Dabasase does not use GUID internally any more since version 1.3.0, so you can set/change it to whatever you like (before that, the cguid was used to map the sourcePackage to a descriptor).

So taking it from there, we have all the freedom to modify the GUID.  

As this is not a catalogue related problem, we should open a corresponding issue in WP3 or even WP2 (not sure if WP4 is using the GUID for anything).

@pchainho could you close this issue here once we have a new issue in WP2/3/4 (wherever it fits best since it is a general architectural discussion) and link to it.

**Comment by pchainho** [comment]
see https://github.com/reTHINK-project/dev-runtime-core/issues/120 for the quick fix and https://github.com/reTHINK-project/specs/issues/12 for the specs related discussion
